sale_block_cancellation/models/sale.py

```diff
@@ -0,0 +1,108 @@
+                        'assigned', 'waiting', 'confirmed', 'done'):
+                    allow = False
+            if not allow or not sale.check_prod_by_locations():
+                sale.allow_cancel = False
+            else:
+                sale.allow_cancel = True
+
+    @api.multi
+    def _get_picking_ids(self):
+        res = super(SaleOrder, self)._get_picking_ids(
+            name='picking_ids', args=None)
+        for sale in self:
+            sale.picking_ids = res.get(sale.id)
+
+    def _search_pickings(self, operator, value):
+        res = []
+        pick_obj = self.env["stock.picking"]
+        pick_ids = []
+        if operator == 'in' and value:
+            pick_ids = pick_obj.browse(value)
+        elif operator and value:
+            pick_ids = pick_obj.search([('name', operator, value)])
+        else:
+            pick_ids = pick_obj.search([])
+        if pick_ids:
+            for pick in pick_ids:
+                proc_gp = pick.group_id
+                if proc_gp:
+                    sale_ids = self.search([
+                        ('procurement_group_id', '=', proc_gp.id)])
+                    res.extend(sale_ids.ids)
+        if operator == '=' and not value:
+            return [('id', 'not in', res)]
+        return [('id', 'in', res)]
+
+    allow_cancel = fields.Boolean(
+        'Allow Cancel', compute=_get_allow_cancel, store=True,
+        help='This field check if all the pickings related to '
+        'this SO are on "Ready to transfer", "Waiting Availability" or in '
+        '"Waiting another move" or if the pickings on different states have '
+        'the move of product quantities correctly returned. If this '
+        'validations are True the sale order cancellation must be allowed.')
+    picking_ids = fields.One2many(
+        compute=_get_picking_ids, method=True, relation='stock.picking',
+        string='Picking associated to this sale', search=_search_pickings)
+
+    @api.multi
+    def check_prod_by_locations(self):
+        self.ensure_one()
+        dict_prod = {}
+        for pick in self.picking_ids:
+            if pick.state == 'done':
+                for move in pick.move_lines:
+                    prod_id = move.product_id.id
+                    loc_s = move.location_id.id
+                    loc_d = move.location_dest_id.id
+                    if prod_id not in dict_prod:
+                        dict_prod.update({prod_id: {}})
+                    qty_s = dict_prod.get(prod_id).get(loc_s, 0.0)
+                    qty_d = dict_prod.get(prod_id).get(loc_d, 0.0)
+                    dict_prod.get(prod_id).update({
+                        loc_s: qty_s - move.product_uom_qty,
+                        loc_d: qty_d + move.product_uom_qty,
+                    })
+        for prod in dict_prod:
+            for loc in dict_prod.get(prod):
+                if dict_prod.get(prod).get(loc) != 0:
+                    return False
+        return True
```

Comment:
Hello @luistorresm 

The main reason for the sql is avoid the excessive use of sentence 'for' in your method.

I was Checking your method and you are using the move to know if the products were correctly returned but with this you need to do something else(a lot of 'for' sentences).

I want to suggest you the use of the quant instead the move, because if the location of the quants used for your moves is different to the origin of the move(It may be stock) you can say that you have products without return.

E.G.
```python
self._cr.execute('''
SELECT product_id, 
       sum(qty),
       location_id 
FROM stock_quant 
WHERE id IN (SELECT quant_id 
             FROM stock_quant_move_rel 
             WHERE move_id IN (SELECT id 
                               FROM stock_move 
                               WHERE picking_id IN %s)) AND
       location_id != %s 
GROUP BY product_id, location_id
''', (tuple(pick_ids.ids), main_loc_id))
```

Where main_loc_id is the location where products initiated the moves.

If this sql sentences returns something this means that you have products without returns, no matter if they return the moves directly from customer to stock.

Do you understand my point?